### PR TITLE
`function` implicit conversion the same way as `fun x`

### PR DESCRIPTION
**Pull request by edgarfgp**
## Description

Fixes https://github.com/dotnet/fsharp/issues/7401

## Checklist

- [x] Test cases added
- [x] Release notes entry updated

### CODE REVIEW

File: src/Compiler/Checking/Expressions/CheckExpressions.fs [hunk 1]
```diff
@@ -9686,6 +9686,8 @@ and GetNewInferenceTypeForMethodArg (cenv: cenv) env tpenv x =
     | SynExpr.Lambda (body = a)
     | SynExpr.DotLambda (expr = a) ->
         mkFunTy g (NewInferenceType g) (GetNewInferenceTypeForMethodArg cenv env tpenv a)
+    | SynExpr.MatchLambda _ ->
+        mkFunTy g (NewInferenceType g) (NewInferenceType g)
```
Comment (brianrourkeboll) [inline src/Compiler/Checking/Expressions/CheckExpressions.fs:9690]: Would it make sense to look at the body of the first branch? Or even to continue looking at the other branches if the first branch leads to `NewInferenceType`?

```suggestion
    | SynExpr.Lambda (body = a)
    | SynExpr.DotLambda (expr = a)
    | SynExpr.MatchLambda (matchClauses = SynMatchClause (resultExpr = a) :: _) ->
        mkFunTy g (NewInferenceType g) (GetNewInferenceTypeForMethodArg cenv env tpenv a)
```

I don't know the answer, but it seems like it _could_ affect inference for something like

```fsharp
M (function x -> function y -> function z -> <@ x + y + z @>)
```

or

```fsharp
M (function
    | None -> fun x -> x + 2
    | Some f -> f)
```

Comment (edgarfgp) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs:9690]: Let’s add some test to find out 👍🏻

**Review by psfinaki — APPROVED** [review]
Nice, thank you Edgar!

File: tests/FSharp.Compiler.ComponentTests/Conformance/BasicGrammarElements/MemberDefinitions/OverloadingMembers/OverloadResolutionUsingFunction.fs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    )
+
+ae.Handle(
+    Func<exn,bool>(
+        function
+        | :? OperationCanceledException -> true
+        | _ -> false        
+    ))
+
+let m () =
```
Comment (abonie) [inline tests/FSharp.Compiler.ComponentTests/Conformance/BasicGrammarElements/MemberDefinitions/OverloadingMembers/OverloadResolutionUsingFunction.fs:22]: I think I am missing the point of this definition here, what is it here for?

Comment (brianrourkeboll) [reply tests/FSharp.Compiler.ComponentTests/Conformance/BasicGrammarElements/MemberDefinitions/OverloadingMembers/OverloadResolutionUsingFunction.fs:22]: I think there should be overloaded methods, at least one of which takes a function. I may have some time today to propose some additional tests.

Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/BasicGrammarElements/MemberDefinitions/OverloadingMembers/OverloadResolutionUsingFunction.fs:22]: Feel free to propose any runnable tests aka that I can copy and paste to this file 🙂

Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/BasicGrammarElements/MemberDefinitions/OverloadingMembers/OverloadResolutionUsingFunction.fs]: @abonie Removed the useless tests. If you have some tests in mind please add these to the conversation. Thanks for the review

**Comment by vzarytovskii** [comment]
> Two comments on the approach used in 581be73:
> 
> 
> 
> 1. I think that it follows the spirit of the existing code in that it simply looks at available syntactic information.
> 
> 2. If we are worried about additional heap allocations (tuples, continuations), we can likely reduce them, perhaps at the expense of some code complexity.

Can we measure the diff between old and new allocation-wise in long-running scenarios?

One-off compilation won't be a concern, continuous allocations in tooling might be. Service currently is not exactly known to be memory-friendly, probably don't want make it worse (I doubt it will be with the change, since nothing will likely escape gen0, but just want a sanity check).

**Comment by brianrourkeboll** [comment]
@vzarytovskii

1. If we [use struct tuples](https://sharplab.io/#v2:DYLgZgzgPg2gPAOQPYGEkFsAOBDATgSwiQDsBuAAmQFEBHAV22HwBcBPCgEQFMxs7hmAQToBzdF2LNszfCQAUvYBC4BKAHwBdALAAoNpi7kAyq2JUAHplzkAvLvIPyUcgAU8E8kjDHTFq+QAqcjpiFkDg0OZwkJZ7R2dBABNE3C4ICAB5by9yACMkJGBwkzNLayCYqIrIuIdnABlsdFzE7E9vfMTWchAfUqtap3IOJGZG5tb28i4ynr6/XEHnAFlpAGMAC3GWtpz0dY2UYD5lCDmS1eZNo5PDJghmJfIARTpRwxzK8PzC4t9Z6phQGPHSOIb1Qq6XTYYiJcjwZBoLB4QgkCjUeiMFjsYY8PgCYRiCRSGTyRTKdQaPqXa7HOjKWzUg43ekfbypCD8ZgLc7/AY6XT6QwAFWFrAMjNF4q4g10wC4UQeuDoa2YwromHlZ3MtkG8qiqTW5GABUwPJm/jWJCivSVKqicgA+uFHSpdaCwY59lcNtNZgB3FgbQZg5wlBYAOjcqWI5Dk2AANORHUmU8m3QBaNQhz0OE1IM2zNpWyQ5urzMoRpIpNKZbxyZjKrhJxPJpPoTPZj25sH5wv+NpyEtROBwchgELkO2quOJLiYZgbJNsTvkOcL30AanIAEZl6wVGWhuHK9tJnJOt0bORsIfu56w3zcBGRmMmjs4xbrNfb+Qs0fe1NHlB2HchR3HSdpwdddF33VcYK3Xc4KPR9+mfGktnfc9vVpW4zmvC5mTpBk5A5LkeR/N0QF6NZiLSTsAMcfVjVNFkGTaa8+2AuNQPAidYyg2d51g8gVz/NQ12ExC91Eg8oXvHtyAAeiU8gISQABrG8okYIpclwGFNjSJMNK4ed8GIERRI2D5iBlBSexU6zpHIER8AANzSYIzkXQx0CQB4p1MKRVXwI0LLAJBcG9WRiAjRiHDkPs2MMVtaLwu9FNDKAJPqQhmAjSLgDhBRJzkOR9nMFLuA3VM3WwM5sDWNY3TkQifRSuMyIECibxUBiHKy8gcN9ZK6JvchA0XBKH3IdLWRqkS2EmmzY3m5RFt9CTKuqqTxLmujNv3GbQ2TfampalCKysCNXneOM0wM/0W1TOrxJm/B2Wwf1rI8UC5AABiTKUDEyobpiUO4gKLHjrTAsd+KnRt7SE2rZPgvbtxklcrudLM5rhwHgbFUH5N7BUkeVGcnQQuDGS42YdXwRJBjYIA===), then I think there should be no additional allocations in the common case (method calls taking anything other than lambdas, address-of, or quotes).
2. Even for method calls taking lambdas, the additional allocations should effectively be equal to the number of curried arguments in the lambda sequence, i.e., one continuation closure each, as in `cont = new loopExpr@24-2<a, b>(cont);`.

**Comment by edgarfgp** [comment]
> @vzarytovskii
> 
> 1. If we [use struct tuples](https://sharplab.io/#v2:DYLgZgzgPg2gPAOQPYGEkFsAOBDATgSwiQDsBuAAmQFEBHAV22HwBcBPCgEQFMxs7hmAQToBzdF2LNszfCQAUvYBC4BKAHwBdALAAoNpi7kAyq2JUAHplzkAvLvIPyUcgAU8E8kjDHTFq+QAqcjpiFkDg0OZwkJZ7R2dBABNE3C4ICAB5by9yACMkJGBwkzNLayCYqIrIuIdnABlsdFzE7E9vfMTWchAfUqtap3IOJGZG5tb28i4ynr6/XEHnAFlpAGMAC3GWtpz0dY2UYD5lCDmS1eZNo5PDJghmJfIARTpRwxzK8PzC4t9Z6phQGPHSOIb1Qq6XTYYiJcjwZBoLB4QgkCjUeiMFjsYY8PgCYRiCRSGTyRTKdQaPqXa7HOjKWzUg43ekfbypCD8ZgLc7/AY6XT6QwAFWFrAMjNF4q4g10wC4UQeuDoa2YwromHlZ3MtkG8qiqTW5GABUwPJm/jWJCivSVKqicgA+uFHSpdaCwY59lcNtNZgB3FgbQZg5wlBYAOjcqWI5Dk2AANORHUmU8m3QBaNQhz0OE1IM2zNpWyQ5urzMoRpIpNKZbxyZjKrhJxPJpPoTPZj25sH5wv+NpyEtROBwchgELkO2quOJLiYZgbJNsTvkOcL30AanIAEZl6wVGWhuHK9tJnJOt0bORsIfu56w3zcBGRmMmjs4xbrNfb+Qs0fe1NHlB2HchR3HSdpwdddF33VcYK3Xc4KPR9+mfGktnfc9vVpW4zmvC5mTpBk5A5LkeR/N0QF6NZiLSTsAMcfVjVNFkGTaa8+2AuNQPAidYyg2d51g8gVz/NQ12ExC91Eg8oXvHtyAAeiU8gISQABrG8okYIpclwGFNjSJMNK4ed8GIERRI2D5iBlBSexU6zpHIER8AANzSYIzkXQx0CQB4p1MKRVXwI0LLAJBcG9WRiAjRiHDkPs2MMVtaLwu9FNDKAJPqQhmAjSLgDhBRJzkOR9nMFLuA3VM3WwM5sDWNY3TkQifRSuMyIECibxUBiHKy8gcN9ZK6JvchA0XBKH3IdLWRqkS2EmmzY3m5RFt9CTKuqqTxLmujNv3GbQ2TfampalCKysCNXneOM0wM/0W1TOrxJm/B2Wwf1rI8UC5AABiTKUDEyobpiUO4gKLHjrTAsd+KnRt7SE2rZPgvbtxklcrudLM5rhwHgbFUH5N7BUkeVGcnQQuDGS42YdXwRJBjYIA===), then I think there should be no additional allocations in the common case (method calls taking anything other than lambdas, address-of, or quotes).
> 2. Even for method calls taking lambdas, the additional allocations should effectively be equal to the number of curried arguments in the lambda sequence, i.e., one continuation closure each, as in `cont = new loopExpr@24-2<a, b>(cont);`.

@vzarytovskii Does Brian's changes cover your expectations ?

**Pull request merged**